## Your submission

You will submit a single `.ipynb` notebook file named with the following convention
```[master]_[name1]_[name2].ipynb```
where 
* `[master]` is any of {BHE, DALETh, EECS}
* `[name1]` and `[name2]` are the names used by you in your academic records, if you have a composed name, then use a dash `-` to separate the parts


In [124]:
""" For Visual Studio Code users : execute this cell first
"""
from pathlib import Path
import os 

old_dir = Path.cwd()  # keep the current directory in memory, may you want to come back to this
new_dir = Path(__vsc_ipynb_file__).parent  # get the directory in which the ipynb file resides
os.chdir(new_dir)  # make the above directory the current working directory --> this will make the below relative import statement correct


### I. start by loading a single paragraph from a chosen lecture from Russell's lectures using the `lectures` returned from `get_lectures`

In [125]:
"""You may need to put russell.py in the same directory as this notebook and import using
from russell import get_lectures
"""
#import the function to 'download' russell lectures
from russell import get_lectures 

#get_lecture give tuple with a dictionnary on the first index, the dictionnaty have one key for each chapter with a dictionnary associated
#in the dictionnary of a chapter there is a 'text' key which contains all the texts of the chapter in a list where each item is a single paragraph
lecture=get_lectures()[0]['I']['text'][2].lower()
#.lower() transform uppercase letter in lowercase letter, if not "The" != 'the' and 'T' != 't' for example

lecture 

'the reasons which i shall give against this theory will be mainly derived from previous authors. there are two sorts of reasons, which will divide my lecture into two parts:'

#### 1. compute the frequency of the letters used in the texts (you may want to use `ascii_lowercase` from the module `string`)

In [126]:
import string 
def letters_occurences(lecture:str) -> dict :
    """
    The argument lecture have to be a string
    The code will return a dictionnary with the occurences of the letters in the string
    
    Examples: 
        letters_occurences('') returns {}
        letters_occurences('ab') returns {'a': 1, 'b': 1}
    """
    #we create a dictionnary for the frequencies of the letters used in the text
    occurences_letter={} 

    for item in lecture: 
        # we check if the character is in the ascii_lowercase to sort letters from non-letters
        if item in string.ascii_lowercase:
            # if the letter don't have an associate key in the dictionnary
            if item not in occurences_letter:
                #we create a key corresponding to the letter and we start the counter with one
                occurences_letter[item]=1
            # if the letter have an associate key in the dictionnary
            else:
                #we add one to the counter associated to the key
                occurences_letter[item]+=1
    return occurences_letter
letters_occurences(lecture)

{'t': 12,
 'h': 10,
 'e': 15,
 'r': 12,
 'a': 9,
 's': 12,
 'o': 11,
 'n': 5,
 'w': 6,
 'i': 14,
 'c': 3,
 'l': 8,
 'g': 2,
 'v': 4,
 'y': 3,
 'b': 1,
 'm': 3,
 'd': 4,
 'f': 2,
 'p': 2,
 'u': 3}

#### 2. compute the frequency of the words in the text(s) (you may want to use `punctuation` from the module `string`, `parse` from the module `russel.py`);

In [129]:
from russell import parse
def take_out_non_letter(lecture:str)-> str:
    """
    The argument lecture have to be a string
    The code will return a string containing only letters and space
    
    Examples: 
        letters_frequency('') returns ''
        letters_frequency('a.b') returns 'ab'
        letters_frequency('a, ui') returns 'ab ui'
    """ 
    #I give to parse a string and the characters I'm looking for inside the given string it gives me back a list of items 
    # a item is either a string containing only letter and space, or a string containing one character of non-letter
    lecture_splitting_punctuation=parse(lecture,string.punctuation) 
    #we gonna build a new string whitout the non-letter
    lecture_whithout_punctuation=''
    #for each element of lecture_splitting_punctuation
    for item in lecture_splitting_punctuation :
        #I check if the element is not non-letter character
        if item not in string.punctuation:
            #I add this element to the new-built string
            lecture_whithout_punctuation+=item
    return lecture_whithout_punctuation

def frequency_words(lecture : str) -> dict :
    """
    The argument lecture have to be a string
    The code will return a dict with the frequency of the words in the string
    
    Examples: 
        letters_frequency('') returns {}
        letters_frequency('ab') returns {'ab' : 1}
        letters_frequency('a, ui') returns {'ab' : 1, 'ui': 1}
    """
    #will return a string containing only letters and space
    cleaned_lecture=take_out_non_letter(lecture)
    #we create a dictionnary for the frequencies of the word used in the text 
    frequencies_word={}
    #splits lecture_whitout_punctuation into a list with the whitespace as separator, each item is a word
    for item in cleaned_lecture.split():
        #I check if word is already registered in frequencies_word dictionnary
        if item not in frequencies_word:
            #if not, I register it and set its counter to one
            frequencies_word[item]=1
        # if yes
        else:
            #I add one to its counter
            frequencies_word[item]+=1
    return frequencies_word   
frequency_words(lecture)

{'the': 1,
 'reasons': 2,
 'which': 2,
 'i': 1,
 'shall': 1,
 'give': 1,
 'against': 1,
 'this': 1,
 'theory': 1,
 'will': 2,
 'be': 1,
 'mainly': 1,
 'derived': 1,
 'from': 1,
 'previous': 1,
 'authors': 1,
 'there': 1,
 'are': 1,
 'two': 2,
 'sorts': 1,
 'of': 1,
 'divide': 1,
 'my': 1,
 'lecture': 1,
 'into': 1,
 'parts': 1}

#### - What happens if words of strictly less than 5 characters are ignored?

In [130]:
#create a dictionnary for words superior or equal to 5 characters
more_five={}
#we go through the words contained in frequency_word
for word in frequency_words(lecture).keys():
    # if the word is superior or equal to 5 characters
    if len(word)>=5:
        #we add it to the dictionnary
        more_five[word]=frequency_words(lecture)[word]
more_five

{'reasons': 2,
 'which': 2,
 'shall': 1,
 'against': 1,
 'theory': 1,
 'mainly': 1,
 'derived': 1,
 'previous': 1,
 'authors': 1,
 'there': 1,
 'sorts': 1,
 'divide': 1,
 'lecture': 1,
 'parts': 1}

In [131]:
#give the keys that are not shared between frequencies_word and frequencies_word_5
frequency_words(lecture).keys()-more_five.keys()

{'are',
 'be',
 'from',
 'give',
 'i',
 'into',
 'my',
 'of',
 'the',
 'this',
 'two',
 'will'}

If words of strictly less than 5 characters are ignored we lost transition words

#### 3. represent word transitions by dictionaries within a dictionary named transitions that will be used as `transitions[word[n]][word[n+1]] = count` where `count` represents the number of occurrences

In [132]:
def word_sequences(lecture:str)-> dict:
    """
    The argument lecture have to be a string
    The code will return dictionnaries whithin a dictionary which represent the word transitions in the string lecture
    
    Examples: 
        word_sequences('') returns {None:{}}
        word_sequences('ab') returns {None:{'ab': 1}, 'ab':{None: 1}}
        word_sequences('a, ui') returns {None:{'a': 1}, 'a':{'ui': 1}, 'ui':{None: 1}}
    """
    #create the dictionnary
    transitions={}
    #create the key None
    transitions[None]={}

    # take out the punctuation and we use split() to make the list iterable
    item=take_out_non_letter(lecture).split()
    #to go through the list
    for k in range(0,len(item)) :
        #if a word do not have an associate key in the dictionnary
        if item[k] not in transitions:
            #we create one and we give it a dictionnary 
            transitions[item[k]]={}
    #to go through the list
    for i in range(0,len(item)):
        #if we are at the beginning at the list, the current word if mandatory following a None 
        if i==0:
            #if the word not have an associate key in the dictionnary transitions[None]
            if item[i] not in transitions[None]:
                #we create one and set its counter to one
                transitions[None][item[i]]=1
            #if yes he does 
            else:
                #we add one to its counter
                transitions[None][item[i]]+=1
        #if we are neither at the beginning nor the end
        if 0<=i<len(item)-1:
            #if the word following liste[i] not have an associate key in the dictionnary transitions[liste[i]]
            if item[i+1] not in transitions[item[i]]:
                #we create one and set its counter to one
                transitions[item[i]][item[i+1]]=1
            #if yes he does
            else:
                #we add one to its counter
                transitions[item[i]][item[i+1]]+=1
        #if we are at the end at the list, the following item if mandatory following a None 
        else:
            #if the word not have an associate key in the dictionnary transitions[liste[len(liste)]]
            if None not in transitions[item[i]]:
                #we create one and set its counter to one
                transitions[item[i]][None]=1
            #if yes he does
            else:
                #we add one to its counter
                transitions[item[i]][None]+=1
    return transitions
word_sequences(lecture)

{None: {'the': 1},
 'the': {'reasons': 1},
 'reasons': {'which': 2},
 'which': {'i': 1, 'will': 1},
 'i': {'shall': 1},
 'shall': {'give': 1},
 'give': {'against': 1},
 'against': {'this': 1},
 'this': {'theory': 1},
 'theory': {'will': 1},
 'will': {'be': 1, 'divide': 1},
 'be': {'mainly': 1},
 'mainly': {'derived': 1},
 'derived': {'from': 1},
 'from': {'previous': 1},
 'previous': {'authors': 1},
 'authors': {'there': 1},
 'there': {'are': 1},
 'are': {'two': 1},
 'two': {'sorts': 1, 'parts': 1},
 'sorts': {'of': 1},
 'of': {'reasons': 1},
 'divide': {'my': 1},
 'my': {'lecture': 1},
 'lecture': {'into': 1},
 'into': {'two': 1},
 'parts': {None: 1}}

#### 4. compute the probability of word transitions Pr(word[n+1] | word[n])

In [133]:
def probability (lecture:str):
    """
    The argument lecture have to be a string
    The code will return the probability of word transitions Pr(word[n+1] | word[n]) based on the given lecture
    
    Examples: 
        word_sequences('') returns None
        word_sequences('ab') returns Pr( a | None )= 1.0 \n Pr( None | a )= 1.0
        word_sequences('a, ui') returns Pr( a | None )= 1.0 \n Pr( ui | a )= 1.0 \n Pr( None | ui )= 1.0
    """
    #assigne the dictionnaries whithin a dictionary which represent the word transitions in the string lecture to a variable
    d=word_sequences(lecture)
    #for each keys items (so the word[n]) included in transitions
    for wordn in d.keys():
        #for each keys items (so the word[n+1]) included in transitions[wordn]
        for wordn1 in d[wordn].keys():
            #the probability is the number of time wordn1 followed wordn divided by the number of time wordn was followed by a word
            prvalue=d[wordn][wordn1]/len(d[wordn1].keys())
            #we print the result
            print("Pr(",wordn1, "|",wordn, ")=", prvalue)
probability(lecture)

Pr( the | None )= 1.0
Pr( reasons | the )= 1.0
Pr( which | reasons )= 1.0
Pr( i | which )= 1.0
Pr( will | which )= 0.5
Pr( shall | i )= 1.0
Pr( give | shall )= 1.0
Pr( against | give )= 1.0
Pr( this | against )= 1.0
Pr( theory | this )= 1.0
Pr( will | theory )= 0.5
Pr( be | will )= 1.0
Pr( divide | will )= 1.0
Pr( mainly | be )= 1.0
Pr( derived | mainly )= 1.0
Pr( from | derived )= 1.0
Pr( previous | from )= 1.0
Pr( authors | previous )= 1.0
Pr( there | authors )= 1.0
Pr( are | there )= 1.0
Pr( two | are )= 0.5
Pr( sorts | two )= 1.0
Pr( parts | two )= 1.0
Pr( of | sorts )= 1.0
Pr( reasons | of )= 1.0
Pr( my | divide )= 1.0
Pr( lecture | my )= 1.0
Pr( into | lecture )= 1.0
Pr( two | into )= 0.5
Pr( None | parts )= 1.0


### II. if all is fool-proof on a single paragragh, generalise all of the above to all of Russell's lectures

In [134]:
from russell import translate_roman
def get_all_russel() -> str :
    """
    This function don't take any argument
    The code will return a string corresponding to all the russel texts
    
    Examples: 
        get_all_russel() returns 'there are...exists.'
    """
    all_text=[]
    all_lectures=''
    for i in range(1,16):
        all_text+=get_lectures()[0][translate_roman(i)]['text']
    for k in range(0,len(all_text)):
        all_lectures+=' '+ all_text[k] #we add a space before the item if not ['...dog']+['The ...] will give ['...dogThe...]
    return all_lectures.lower()
all_lecture=get_all_russel()
all_lecture

' there are certain occurrences which we are in the habit of calling "mental." among these we may take as typical believing and desiring. the exact definition of the word "mental" will, i hope, emerge as the lectures proceed; for the present, i shall mean by it whatever occurrences would commonly be called mental. i wish in these lectures to analyse as fully as i can what it is that really takes place when we, e.g. believe or desire. in this first lecture i shall be concerned to refute a theory which is widely held, and which i formerly held myself: the theory that the essence of everything mental is a certain quite peculiar something called "consciousness," conceived either as a relation to objects, or as a pervading quality of psychical phenomena. the reasons which i shall give against this theory will be mainly derived from previous authors. there are two sorts of reasons, which will divide my lecture into two parts: (1) direct reasons, derived from analysis and its difficulties; (2

In [135]:
from russell import parse
def text_cleaner(lecture:str) -> str:
    """
    The argument have to be a string
    The code will return a string containing only word, dealing with all special cases
    
    Examples:
        text_cleaner('') returns ''
        text_cleaner('According to the chapter alpha (p.24 ii)') returns 'According to the chapter alpha'
    """
    #I give to parse a string and the characters I'm looking for inside the given string and it t gives me back a list of items
    #an item is either a string containing only letter and space, or a string containing one character researched
    uncleaned_lecture=parse(lecture, string.punctuation +'0123456789'+'—')
    #create an empty string to build a string containing only word
    cleaned_lecture=''
    #create a parameter to know if the item we are looking at is between parenthesis or not
    into_parenthesis=False

    #we are going trough every item in uncleaned_lecture
    for item in uncleaned_lecture: 
        #parenthesis is opened
        if item =='(':
           into_parenthesis=True
        #parenthesis is closed
        if item==')':
            into_parenthesis=False
        #check if the item is not a researched character and if it's not between parenthensis
        if item not in string.punctuation+'0123456789'+'—' and into_parenthesis==False:
            cleaned_lecture +=item 
    return cleaned_lecture
text_cleaner(all_lecture)

' there are certain occurrences which we are in the habit of calling mental among these we may take as typical believing and desiring the exact definition of the word mental will i hope emerge as the lectures proceed for the present i shall mean by it whatever occurrences would commonly be called mental i wish in these lectures to analyse as fully as i can what it is that really takes place when we eg believe or desire in this first lecture i shall be concerned to refute a theory which is widely held and which i formerly held myself the theory that the essence of everything mental is a certain quite peculiar something called consciousness conceived either as a relation to objects or as a pervading quality of psychical phenomena the reasons which i shall give against this theory will be mainly derived from previous authors there are two sorts of reasons which will divide my lecture into two parts  direct reasons derived from analysis and its difficulties  indirect reasons derived from o

#### 1. display the frequencies of the letters used in the text and search for the relative occurrences of letters in English texts, compare and conclude

In [ ]:
def letters_frequency(lecture:str) -> dict:
    """
    The argument have to be a string
    The code will return the frequency of letter used in the string
    Examples:
        letters_frequency('') returns {}
        letters_frequency('abbb') returns {'a': 25.0, 'b': 75.0}
    """
    #
    d={}
    d['count']=letters_occurences(text_cleaner(lecture))
    d['frequency']={}
    addition=0
    for item in d['count'].values():
        addition+=item
    for letter in d['count'].keys():
        d['frequency'][letter]=d['count'][letter]/addition*100
    return d['frequency']
letters_frequency(all_lecture)

{'t': 9.610552923071463,
 'h': 5.269869761758077,
 'e': 12.900608724973441,
 'r': 5.360379889004667,
 'a': 7.748630305427312,
 'c': 3.7633958791271476,
 'i': 8.056161914251161,
 'n': 7.0237886981499935,
 'o': 7.499410543288939,
 'u': 2.633413702269852,
 's': 7.305713800385873,
 'w': 2.006688115285056,
 'b': 1.675578238018817,
 'f': 2.532762384295352,
 'l': 3.6102638991357168,
 'g': 1.6829306012965513,
 'm': 2.5515235871419844,
 'y': 1.852034956684439,
 'k': 0.40894351610294327,
 'p': 2.021392841840525,
 'v': 1.0288237993464002,
 'd': 2.8428292908011796,
 'x': 0.2717839115079697,
 'q': 0.13082136039003017,
 'j': 0.17138612330166747,
 'z': 0.04031123314343953}

|Letter |	Count	| 	Letter	|Frequency|
|---    |:-:    |:-:    |:-:    |
|E|	21912|	 	E|	12.02|
|T|	16587|	 	T|	9.10|
|A|	14810|	 	A|	8.12|
|O|	14003|	 	O|	7.68|
|I|	13318|	 	I|	7.31|
|N|	12666|	 	N|	6.95|
|S|	11450|	 	S|	6.28|
|R|	10977|	 	R|	6.02|
|H|	10795|	 	H|	5.92|
|D|	7874|	 	D|	4.32|
|L|	7253|	 	L|	3.98|
|U|	5246|	 	U|	2.88|
|C|	4943|	 	C|	2.71|
|M|	4761|	 	M|	2.61|
|F|	4200|	 	F|	2.30|
|Y|	3853|	 	Y|	2.11|
|W|	3819|	 	W|	2.09|
|G|	3693|	 	G|	2.03|
|P|	3316|	 	P|	1.82|
|B|	2715|	 	B|	1.49|
|V|	2019|	 	V|	1.11|
|K|	1257|	 	K|	0.69|
|X|	315	 |	X|	0.17|
|Q|	205	 |	Q|	0.11|
|J|	188	 |	J|	0.10|
|Z|	128	 |	Z|	0.07|


I didn't check for every letters, but the frequency for each letter used in the text is close to the relative occurrences of letters in English texts. I think we can conclude that russel text is a good example to train a future model to make generative text whitout having our model to train itself over all the english texts.

#### 2. display the frequencies of the words used in the text, before and after having filtered out the `most_frequent_words`. What do these most frequent words say about the topic of the lectures?

In [111]:
fq_word=frequency_words(all_russel_cleaned)
fq_word

{'there': 355,
 'are': 940,
 'certain': 148,
 'occurrences': 64,
 'which': 1028,
 'we': 1298,
 'in': 1996,
 'the': 5296,
 'habit': 81,
 'of': 3979,
 'calling': 8,
 'mental': 131,
 'among': 38,
 'these': 150,
 'may': 442,
 'take': 61,
 'as': 1006,
 'typical': 4,
 'believing': 30,
 'and': 1595,
 'desiring': 4,
 'exact': 11,
 'definition': 74,
 'word': 189,
 'will': 274,
 'i': 575,
 'hope': 7,
 'emerge': 1,
 'lectures': 17,
 'proceed': 11,
 'for': 435,
 'present': 165,
 'shall': 114,
 'mean': 52,
 'by': 684,
 'it': 1328,
 'whatever': 29,
 'would': 210,
 'commonly': 17,
 'be': 1066,
 'called': 124,
 'wish': 49,
 'to': 2477,
 'analyse': 10,
 'fully': 6,
 'can': 307,
 'what': 441,
 'is': 2402,
 'that': 1660,
 'really': 42,
 'takes': 9,
 'place': 94,
 'when': 362,
 'eg': 13,
 'believe': 98,
 'or': 565,
 'desire': 125,
 'this': 782,
 'first': 133,
 'lecture': 73,
 'concerned': 60,
 'refute': 2,
 'a': 2116,
 'theory': 86,
 'widely': 6,
 'held': 18,
 'formerly': 5,
 'myself': 16,
 'essence': 20,

Most frequent words as found in a wikipedia page on [Most common words in English](https://en.wikipedia.org/wiki/Most_common_words_in_English), consulted on 2020-09-13.

In [112]:
most_frequent_words = {
    'the', 'be', 'are', 'to', 'of', 'and', 'a', 'in', 'that', 'have', 'I','i',
    'it', 'for', 'not', 'on', 'with', 'he', 'as', 'you', 'do',
    'at', 'this', 'but', 'his', 'by', 'from', 'they', 'we', 'say',
    'her', 'she', 'or', 'an', 'will', 'my', 'one', 'all', 'would',
    'there', 'their', 'what', 'so', 'up', 'out', 'if', 'about',
    'who', 'get', 'which', 'go', 'me', 'when', 'make', 'can', 
    'like', 'time', 'no', 'just', 'him', 'know', 'take', 'people', 
    'into', 'year', 'good', 'some', 'could', 'them', 'see', 'other', 
    'than', 'then', 'now', 'look', 'only', 'come', 'its', 'over', 
    'think', 'also', 'back', 'after', 'use', 'two', 'how', 'our', 
    'work', 'first', 'well', 'way', 'even', 'new', 'want', 'because', 
    'any', 'these', 'give', 'day', 'most', 'us'}
# add different tenses and times of the verbs
most_frequent_words.update({'am','is', 'are', 'was', 'were', 'has', 'have', 'had', 'did', 'done', 'said', 'made', 
                            'could', 'took', 'taken', 'saw', 'seen', 'looked', 'worked', '1', '2', 'wanted', 
                            'gave', 'given','got','went','gone','knew','known','came','thought','used'})
# exclude words with a possible physical interpretation
most_frequent_words.difference_update({'time', 'work'})

In [113]:
def frequency_words_filtered(d:dict)-> dict:
    filtered={}
    for item in d.keys():
        if item not in most_frequent_words:
            filtered[item]=d[item]
    return filtered
fq_word_filtered=frequency_words_filtered(fq_word)
fq_word_filtered

{'certain': 148,
 'occurrences': 64,
 'habit': 81,
 'calling': 8,
 'mental': 131,
 'among': 38,
 'may': 442,
 'typical': 4,
 'believing': 30,
 'desiring': 4,
 'exact': 11,
 'definition': 74,
 'word': 189,
 'hope': 7,
 'emerge': 1,
 'lectures': 17,
 'proceed': 11,
 'present': 165,
 'shall': 114,
 'mean': 52,
 'whatever': 29,
 'commonly': 17,
 'called': 124,
 'wish': 49,
 'analyse': 10,
 'fully': 6,
 'really': 42,
 'takes': 9,
 'place': 94,
 'eg': 13,
 'believe': 98,
 'desire': 125,
 'lecture': 73,
 'concerned': 60,
 'refute': 2,
 'theory': 86,
 'widely': 6,
 'held': 18,
 'formerly': 5,
 'myself': 16,
 'essence': 20,
 'everything': 22,
 'quite': 42,
 'peculiar': 10,
 'something': 115,
 'consciousness': 98,
 'conceived': 3,
 'either': 56,
 'relation': 99,
 'objects': 93,
 'pervading': 1,
 'quality': 9,
 'psychical': 20,
 'phenomena': 77,
 'reasons': 18,
 'against': 31,
 'mainly': 4,
 'derived': 24,
 'previous': 22,
 'authors': 4,
 'sorts': 20,
 'divide': 3,
 'parts': 32,
 'direct': 15,
 '

In [114]:
sorted_items = sorted(fq_word_filtered.items(), key=lambda kv: (kv[1], kv[0]))

sorted_dict={}
for item in sorted_items[::-1]:
    sorted_dict[item[0]]=item[1]
sorted_dict

{'may': 442,
 'images': 278,
 'such': 251,
 'words': 227,
 'more': 220,
 'sensations': 202,
 'belief': 198,
 'case': 195,
 'image': 194,
 'different': 192,
 'word': 189,
 'knowledge': 179,
 'same': 176,
 'object': 173,
 'present': 165,
 'must': 165,
 'view': 158,
 'fact': 158,
 'between': 158,
 'laws': 154,
 'memory': 150,
 'certain': 148,
 'very': 146,
 'physical': 144,
 'experience': 138,
 'question': 136,
 'matter': 136,
 'past': 133,
 'sensation': 132,
 'been': 132,
 'those': 131,
 'mental': 131,
 'desire': 125,
 'called': 124,
 'does': 123,
 'cannot': 120,
 'something': 115,
 'shall': 114,
 'mind': 113,
 'without': 110,
 'true': 110,
 'psychology': 108,
 'content': 108,
 'upon': 106,
 'sense': 106,
 'meaning': 105,
 'therefore': 104,
 'part': 103,
 'movements': 100,
 'relation': 99,
 'being': 99,
 'merely': 98,
 'consciousness': 98,
 'believe': 98,
 'general': 97,
 'cause': 97,
 'things': 96,
 'beliefs': 96,
 'world': 94,
 'place': 94,
 'physics': 93,
 'objects': 93,
 'causal': 93

#### What do these most frequent words say about the topic of the lectures?

These frequent words say that the topic is related to mind, psychology, memory,... So the topic must Cognitive science.

#### 3. compute the word transitions dictionary

In [115]:
word_sequences=word_sequences(all_russel_cleaned)
word_sequences

{None: {'there': 1},
 'there': {'are': 70,
  'is': 186,
  'must': 9,
  'can': 9,
  'seem': 1,
  'fore': 1,
  'will': 9,
  'would': 5,
  'may': 13,
  'seems': 3,
  'has': 3,
  'after': 1,
  'need': 2,
  'being': 1,
  'was': 2,
  'any': 2,
  'were': 7,
  'which': 1,
  'certainly': 4,
  'remain': 1,
  'had': 1,
  'remains': 3,
  'and': 1,
  'have': 1,
  'really': 1,
  'could': 3,
  'thats': 1,
  'when': 1,
  'the': 2,
  'be': 1,
  'used': 1,
  'adds': 1,
  'also': 1,
  'judging': 1,
  'an': 1,
  'should': 1,
  'cannot': 1,
  'defined': 1,
  'decided': 1,
  'ultimately': 1},
 'are': {'certain': 5,
  'in': 30,
  'two': 5,
  'more': 9,
  'not': 76,
  'willing': 1,
  'called': 11,
  'compounded': 2,
  'conscious': 7,
  'aware': 5,
  'right': 6,
  'answered': 1,
  'possessed': 1,
  'various': 3,
  'opposed': 1,
  'other': 2,
  'those': 12,
  'all': 10,
  'to': 27,
  'interested': 1,
  'phenomena': 2,
  'vast': 1,
  'three': 1,
  'thinking': 4,
  'necessarily': 1,
  'merely': 8,
  'outside': 1,

### III. make your own Russell sentences based on the computed `transitions` where the sentence would be constructed by _sampling_ a next word based on the relative counts given the current word (see also the caseine exercise as an example)

In [116]:
import random    
rw = ''
last_char = None
while True:  # draw a random sequence of characters, starting with None
    c_ = random.choices(list(word_sequences[last_char].keys()),list(word_sequences[last_char].values()))  #random.choices(list of letter,list of theirs associates weight) -> random.choice(['a','b'],[2,1]), je tirerai 2 fois plus de a que de b
    #list(char_sequences[last_char].keys()) renvoie les clés du dictionnaire, quelles lettres sont couramment utilisés après la derniere lettre qu'on vient d'utiliser selon le dictionnaire char_sequences
    #list(char_sequences[last_char].values()) donne la liste des poids associés
    c = c_[0]  # since we ask only for a single element, we need to pick the first and only element from the list
    if c is None:  # means we're at the end of a word
        break
    rw += c +' '  # add character c as the next character to the word
    last_char = c
    
print(rw)

KeyboardInterrupt: 

To go further (bonus):
* use a class that represents the graph with a method `step` (go from word[n] to word[n+1]) and `walk` (create a pseudo-sentence)

In [ ]:
import random   
class sentence (object) :
    def __init__(self,word_sequences):
        self.word=None
        self.pseudo_sentence=''
        self.word_sequences=word_sequences
        self.step()
    def step(self):
        while True:
            self.word=random.choices(list(self.word_sequences[self.word].keys()),list(self.word_sequences[self.word].values()))[0]
            if self.word == None :
                break
            self.walk()
    def walk(self):
        self.pseudo_sentence+=self.word+' '

if __name__ == "__main__": 
    sentence(word_sequences)
sentence(word_sequences).pseudo_sentence

'there can theoretically as to the reference to be able to hear a sheep or any rate prima facie very little relation of us to show that it was that only remembering one actor we are unusually reflective judgment which are still influential and two books mentioned in this occurred in vacuo together with this fact that if we have now by defects in the past event but merely a certain feature in the word has been right place and place but also to whether in the patch of particulars which do not only takes it is usually lost in explaining the general and pleasure as an inference is to consider the process which differ more severe critics of things would i merely the belief may be abolished nor subjective certainty can distinguish perceptions from this school the argument could be confident i think that such claim except when an external unscientific since no movement which is not injured by which we ought certainly is easy when he gives knowledge is conscious desire which may be contended t